plugins/c-ares-dns-client.c

```diff
@@ -0,0 +1,757 @@
+			       gboolean destroy_old_source);
+static void stop_ares_request(struct ares_request *request);
+
+/*
+ * Set/unset reverse path filtering (from connman). Using this should make
+ * setting routes for the DNS servers unnecessary, but unfortunately that is not
+ * happening.
+ * TODO Investigate why this is not working.
+ */
+static void rp_filter_set(const char *interface, gboolean enabled)
+{
+	int fd;
+	ssize_t cnt;
+	char filename[PATH_MAX];
+	const char *str_value = (enabled == TRUE) ? "1" : "0";
+
+	snprintf(filename, sizeof(filename),
+			"/proc/sys/net/ipv4/conf/%s/rp_filter", interface);
+	fd = open(filename, O_WRONLY);
+	if (fd == -1)
+		return;
+	cnt = write(fd, str_value, strlen(str_value));
+	if (cnt == -1)
+		ofono_error("%s: cannot write (%s)", __func__, strerror(errno));
+	close(fd);
+}
+
+/*
+ * Callback invoked when it's time to give control back to c-ares. Controlled by
+ * the glib source referred to by |timeout_source_id| in struct ares_request.
+ */
+static gboolean ares_timeout_cb(gpointer data)
+{
+	struct ares_request *request = data;
+	const gboolean destroy_old_source = FALSE;
+
+	DBG("request %p: running = %d", request, request->running);
+
+	if (!request->running) {
+		request->timeout_source_id = 0;
+		return FALSE;
+	}
+
+	ares_process_fd(request->channel, ARES_SOCKET_BAD, ARES_SOCKET_BAD);
+
+	/*
+	 * NOTE: We tell reset_ares_timeout not to destroy its old timer source
+	 * because we're calling it from within that source and it will be
+	 * destroyed by glib when we return FALSE below.
+	 */
+	reset_ares_timeout(request, destroy_old_source);
+
+	/*
+	 * Return FALSE to get rid of our old glib source. We created a new
+	 * one during our call to reset_ares_timeout above.
+	 */
+	return FALSE;
+}
+
+/*
+ * Determine how long c-ares is willing to wait until being given control and
+ * schedule ares_timeout_cb to be invoked at that time. Any existing
+ * timer is replaced. If |destroy_old_source| is TRUE, the old timer's glib
+ * source will be destroyed.
+ */
+static void reset_ares_timeout(struct ares_request *request,
+			       gboolean destroy_old_source)
+{
+	struct timeval ret_tv, now, elapsed, max_tv;
+	struct timeval *tv;
+	struct timeval *max = NULL;
+	guint timeout_interval_msecs = 0;
+	gboolean timeout_provided = FALSE;
+
+	DBG("request %p: running = %d", request, request->running);
+
+	if (!request->running)
+		return;
+
+	/*
+	 * Compute how much time has elapsed since the request started.
+	 * If the client provided a non-default timeout and we've timed out,
+	 * notify the client and stop the request.
+	 */
+	gettimeofday(&now, NULL);
+	timersub(&now, &request->start_time, &elapsed);
+	timeout_provided = request->timeout.tv_sec != 0 ||
+			   request->timeout.tv_usec != 0;
+	if (timeout_provided && timercmp(&elapsed, &request->timeout, >=)) {
+		request->cb(request->data, OFONO_DNS_CLIENT_ERROR_TIMED_OUT,
+			    NULL);
+		stop_ares_request(request);
```

Comment:
It seems odd that the request is stopped *after* the timeout callback in invoked...